### DISCUSSION TITLE
Invalid code behaviour after babel apply

### DISCUSSION THREAD

**Discussion by Maximys** [babel/babel#16574] [Q&A] [answered]
Hello everybody.
Today I had spend whole working day for understand why my extremely simple code works incorrect. For provide some context, I'll try to explain what we are using and what's happened.
We have ASP.NET Core application with frontend, which use TypeScript. For providing REST API Clients to frontend, we use [NSwag](https://github.com/RicoSuter/NSwag). Also we use babel.
I should implemented some new API endpoints, which result should be displayed on frontend. Some of this endpoints have the similar model, that's why I created `EndpointBaseDto` and `Endpoint1Dto`, `Endpoint2Dto`, which inherited from `EndpointBaseDto`. All TypeScript code is very simple:

1. DTO initialization by code like:
```
let myDto1 = new Endpoint1Dto({
propertyFromBase1 = valueForPropertyFromBase1,
propertyFromBase2 = valueForPropertyFromBase2,
...,
propertyFromBaseN = valueForPropertyFromBaseN,
propertyFromEndpoint1 = valueForPropertyFromEndpoint1
})
```

2.call ApiClient with `myDto1`.

But my API does not receive  `propertyFromEndpoint1` . First of all, I set breakpoint to line where `myDto1` was initialized and check `valueForPropertyFromEndpoint1` - everything was fine. Then I set breakpoint in the `Endpoint1Dto` class constructor inside TypeScript file, which was generated by NSwag and I found that `propertyFromEndpoint1` has `undefined` value. Then I set breakpoint in the `EndpointBaseDto` class constructor, which should fill all properties and `propertyFromEndpoint1` was initialized inside it. It's mean, that some other code reset `propertyFromEndpoint1`. Then I had check JS-file, generated by translation of my TypeScript code and I found next lines of JS-code:
```
...
super(data);
this.propertyFromEndpoint1 = void 0;
...
```
After that I start to Google some info about this interesting bug and found next link: [Class property type definition is compiled to assigning void 0](https://github.com/babel/babel/issues/8280), where @nicolo-ribaudo wrote:

> It's correct, class fields are defined after the super constructor has been called. I think that declaring the field like foo: type = this.foo might work.

Ok. Next I looks to NSwag repo and found bug, which was the same as my: [TypeScript inheritance constructors is not created correctly ](https://github.com/RicoSuter/NSwag/issues/3580), but I agree with @RicoSuter [here](https://github.com/RicoSuter/NJsonSchema/pull/1485#issuecomment-1027006183), that's it's not bug of NSwag(I hope, everybody of you will agree with me, that if after applying some obfuscators your code starting to work invalid, then it's bug of obfuscators and not of your code).

Could anybody of yours looks to this problem and reply me am I and @RicoSuter correct or not? And if we are correct, could you try to fix this bug? If you have no time for it, I can try to fix it by myself and create PR to current repository

**Comment by liuxingbaoyu** [comment]
Babel matches the behavior with this option enabled.
https://www.typescriptlang.org/tsconfig/#useDefineForClassFields

https://www.typescriptlang.org/play/?noImplicitAny=false&useDefineForClassFields=true&target=1&ts=5.5.1-rc&experimentalDecorators=false#code/MYGwhgzhAEDC0G8CwAoa7pgOYFMBc0AdgK4C2ARjgE4A+xhAJjgGYCWhODqAvkA

**Reply by JLHwung** [reply]
Note that the typescript option `useDefineForClassFields` defaults to `true` when targeting `ES2022`, which means TypeScript already aligns to Babel's behaviour for recent targets. Since ES2022 has been supported by recent versions of all major browsers, I think NSwag should adapt to the spec as well.

**Reply by Maximys** [reply]
By my opinion, babel should apply adding property initialization by default value at the beginning of constructor and not after `super()` call. I mean, that JS code should be the next;
```
...
this.propertyFromEndpoint1 = void 0;
super(data);
...
```

**Reply by liuxingbaoyu** [reply, marked as answer]
https://www.typescriptlang.org/play/?noImplicitAny=false&useDefineForClassFields=true&target=1&noCheck=true&ts=5.5.1-rc&experimentalDecorators=false#code/MYGwhgzhAEDC0FMAeAXBA7AJjWAmaA3gLABQ050YA5ggFzToCuAtgEYIBOAPo1ggGYBLdAkykK0YAHt0EFB0bAUUjgAoAlMTISKERgAdOGgNziKAX1LmgA
TS has the same behavior.

**Reply by Maximys** [reply]
Thanks, I had catch your info
